FeedStoreChallenge/CoreData/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,54 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				if let cache = try ManagedCache.find(in: context) {
+					completion(.found(feed: cache.localFeed,
+					                  timestamp: cache.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				let feed = ManagedFeedImage.images(from: feed, in: context)
+				let cache = try ManagedCache.newUniqueInstance(in: context)
+				cache.timestamp = timestamp
+				cache.feed = feed
+
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
 	}
 
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				try ManagedCache.find(in: context).map(context.delete)
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
+	}
+
+	func perform(_ action: @escaping (NSManagedObjectContext) -> Void) {
+		let context = self.context
+		context.perform {
+			action(context)
+		}
```

Comment:
Suggestion:
During the course this way of having a local reference to the `context` to avoid capturing `self` inside of the `context.perform` block is taught.
The same can be achieved using a capture list.
It saves one line and one variable.

No need to make this amendment, this is just for your interest.
```suggestion
		context.perform { [context] in 
			action(context) 
		}
```